File: Step03_Arrays/Lec1_Easy/Program_06_better.cpp
#include <bits/stdc++.h>
using namespace std;

// Left Rotate an array by 'd' places

int main() {
    int n, d;
    cin >> n;
    cin >> d;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    d = d % n;
    vector<int> temp;
    for (int i = 0; i < d; i++) {
        temp.push_back(arr[i]);
    }
    for (int i = d; i < n; i++) {
        arr[i - d] = arr[i];
    }
    for (int i = n - d; i < n; i++) {
        arr[i] = temp[i - (n - d)];
    }

    for (auto element : arr) {
        cout << element << " ";
    }
}

// Time-complexity:  O(n+d)
// Space-complexity: O(d)

File: Step03_Arrays/Lec1_Easy/Program_06_better.java
package Step03_Arrays.Lec1_Easy;

import java.util.*;

// Left Rotate an array by 'd' places

public class Program_06_better {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int n = in.nextInt();
        int d = in.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            arr.add(in.nextInt());
        }

        d = d % n;
        ArrayList<Integer> temp = new ArrayList<>();
        for (int i = 0; i < d; i++) {
            temp.add(arr.get(i));
        }
        for (int i = d; i < n; i++) {
            arr.set(i - d, arr.get(i));
        }
        for (int i = n - d; i < n; i++) {
            arr.set(i, temp.get(i - (n - d)));
        }

        for (var element : arr) {
            System.out.print(element + " ");
        }

        in.close();
    }
}

// Time-complexity : O(n+d)
// Space-complexity : O(d)
